rest/oidc_api.go

```diff
@@ -103,114 +101,123 @@ func (h *handler) handleOIDCCommon() (redirectURLString string, err error) {
 }
 
 func (h *handler) handleOIDCCallback() error {
-	callbackError := h.getQuery("error")
+	callbackError := h.getQuery(requestParamError)
 	if callbackError != "" {
-		errorDescription := h.getQuery("error_description")
-		return base.HTTPErrorf(http.StatusUnauthorized, "oidc_callback received an error: %v", errorDescription)
+		errorDescription := h.getQuery(requestParamErrorDesc)
+		return base.HTTPErrorf(http.StatusUnauthorized, "oidc callback received an error: %v", errorDescription)
 	}
 
-	code := h.getQuery("code")
+	code := h.getQuery(requestParamCode)
 	if code == "" {
 		return base.HTTPErrorf(http.StatusBadRequest, "Code must be present on oidc callback")
 	}
 
-	providerName := h.getQuery("provider")
+	providerName := h.getQuery(requestParamProvider)
 	provider, err := h.getOIDCProvider(providerName)
 	if err != nil || provider == nil {
 		return base.HTTPErrorf(http.StatusBadRequest, "Unable to identify provider for callback request")
 	}
 
-	oac, err := provider.GetClient(h.getOIDCCallbackURL).OAuthClient()
-	if err != nil {
+	client := provider.GetClient(h.getOIDCCallbackURL)
+	if client == nil {
 		return err
 	}
 
-	tokenResponse, err := oac.RequestToken(oauth2.GrantTypeAuthCode, code)
+	// Converts the authorization code into a token.
+	token, err := client.Config.Exchange(context.Background(), code)
 	if err != nil {
-		return err
+		return base.HTTPErrorf(http.StatusUnauthorized, "Failed to exchange token: "+err.Error())
+	}
+
+	rawIDToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		return base.HTTPErrorf(http.StatusInternalServerError, "No id_token field in oauth2 token.")
```

Comment:
Why do you think we should return a StatusInternalServerError here, as opposed to StatusUnauthorized as we're doing above on the failed exchange?